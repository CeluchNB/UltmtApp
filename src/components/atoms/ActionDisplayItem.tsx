import { GuestTeam } from '../../types/team'
import React from 'react'
import { useTheme } from '../../hooks'
import {
    Action,
    LiveServerActionData,
    SavedServerActionData,
    ServerActionData,
} from '../../types/action'
import { Chip, IconButton } from 'react-native-paper'
import { Pressable, StyleSheet, Text, View } from 'react-native'

interface ActionDisplayItemProps {
    action: Action
    teamOne: GuestTeam
    teamTwo: GuestTeam
    onPress?: (action: ServerActionData) => void
}

const ActionDisplayItem: React.FC<ActionDisplayItemProps> = ({
    action,
    teamOne,
    teamTwo,
    onPress,
}) => {
    const {
        theme: { colors, size, weight },
    } = useTheme()
    const { action: actionData, viewerDisplay } = action
    const { tags, comments } = actionData

    const getTeamName = (): string => {
        if ((actionData as any).teamNumber) {
            return (actionData as LiveServerActionData).teamNumber === 'one'
                ? teamOne.name
                : teamTwo.name
        } else if ((actionData as any).team) {
            return (actionData as SavedServerActionData).team.name
        }
        return teamOne.name
    }

    const styles = StyleSheet.create({
        container: {
            backgroundColor: colors.darkPrimary,
            shadowColor: colors.darkGray,
            shadowRadius: 8,
            borderRadius: 8,
            width: '100%',
            padding: 10,
        },
        dataContainer: {
            display: 'flex',
            flexDirection: 'row',
        },
        textContainer: {
            flex: 1,
        },
        lineContainer: {
            display: 'flex',
            flexDirection: 'row',
            flexWrap: 'wrap',
            alignItems: 'center',
        },
        team: {
            color: colors.textSecondary,
            fontWeight: weight.full,
            fontSize: size.fontFifteen,
        },
        action: {
            color: colors.textPrimary,
            fontWeight: weight.bold,
            fontSize: size.fontFifteen,
        },
        chipContainer: {
            flex: 1,
            flexDirection: 'row',
        },
        chip: {
            borderRadius: 4,
            margin: 2,
            backgroundColor: colors.primary,
            borderWidth: 1,
        },
        commentCount: {
            color: colors.textPrimary,
            fontSize: size.fontFifteen,
        },
    })

    return (
        <Pressable
            style={styles.container}
            disabled={!onPress}
            android_ripple={{ color: colors.textPrimary }}
            onPress={() => {
                if (onPress) {
                    onPress(actionData)
                }
            }}>
            <Text style={styles.team}>{getTeamName()}</Text>
            <View style={styles.dataContainer}>
                <View style={styles.textContainer}>
                    <View style={styles.lineContainer}>
                        <View>
                            <Text style={styles.action}>{viewerDisplay}</Text>
                        </View>
                    </View>
                    <View style={styles.lineContainer}>
                        {tags.map(tag => (
                            <Chip
                                key={tag}
                                style={styles.chip}
                                mode="outlined"
                                selectedColor={colors.textPrimary}
                                ellipsizeMode="tail">
                                {tag}
                            </Chip>
                        ))}
                        <IconButton
                            size={15}
                            icon="comment-text-outline"
                            iconColor={colors.textPrimary}
                        />
                        {comments.length > 0 && (
                            <Text style={styles.commentCount}>
                                ({comments.length})
                            </Text>
                        )}
                    </View>
                </View>
            </View>
        </Pressable>
    )
}

export default ActionDisplayItem
