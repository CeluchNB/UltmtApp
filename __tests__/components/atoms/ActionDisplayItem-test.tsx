import ActionDisplayItem from '../../../src/components/atoms/ActionDisplayItem'
import React from 'react'
import {
    ActionFactory,
    ActionType,
    LiveServerActionData,
    SavedServerActionData,
} from '../../../src/types/action'
import { fireEvent, render } from '@testing-library/react-native'

describe('ActionDisplayItem', () => {
    it('should match snapshot with server action', () => {
        const snapshot = render(
            <ActionDisplayItem
                action={ActionFactory.createFromAction({
                    _id: '',
                    actionType: ActionType.CATCH,
                    actionNumber: 1,
                    playerOne: { firstName: 'Noah', lastName: 'Celuch' },
                    playerTwo: {
                        firstName: 'Connor',
                        lastName: 'Tipping',
                    },
                    tags: ['break', 'huck'],
                    comments: [
                        {
                            user: {
                                firstName: 'First',
                                lastName: 'last',
                            },
                            comment: 'nice',
                            commentNumber: 1,
                        },
                    ],
                    team: {
                        _id: '',
                        place: 'Pittsburgh',
                        name: 'Temper',
                        teamname: 'temper',
                        seasonStart: '2022',
                        seasonEnd: '2022',
                    },
                } as SavedServerActionData)}
                teamOne={{ name: 'Team 1' }}
                teamTwo={{ name: 'Team 2' }}
            />,
        )
        expect(snapshot.toJSON()).toMatchSnapshot()
    })

    it('match snapshot with live action', () => {
        const snapshot = render(
            <ActionDisplayItem
                action={ActionFactory.createFromAction({
                    actionType: ActionType.CATCH,
                    actionNumber: 1,
                    playerOne: { firstName: 'Noah', lastName: 'Celuch' },
                    playerTwo: {
                        firstName: 'Connor',
                        lastName: 'Tipping',
                    },
                    tags: ['break', 'huck'],
                    comments: [
                        {
                            user: {
                                firstName: 'First',
                                lastName: 'last',
                            },
                            comment: 'nice',
                            commentNumber: 1,
                        },
                    ],
                    teamNumber: 'two',
                } as LiveServerActionData)}
                teamOne={{ name: 'Team 1' }}
                teamTwo={{ name: 'Team 2' }}
            />,
        )
        expect(snapshot.toJSON()).toMatchSnapshot()
    })

    it('should handle press', () => {
        const onPress = jest.fn()

        const { getByText } = render(
            <ActionDisplayItem
                action={ActionFactory.createFromAction({
                    _id: '',
                    actionType: ActionType.CATCH,
                    actionNumber: 1,
                    playerOne: { firstName: 'Noah', lastName: 'Celuch' },
                    playerTwo: {
                        firstName: 'Connor',
                        lastName: 'Tipping',
                    },
                    tags: ['break', 'huck'],
                    comments: [
                        {
                            user: {
                                firstName: 'First',
                                lastName: 'last',
                            },
                            comment: 'nice',
                            commentNumber: 1,
                        },
                    ],
                    team: {
                        _id: '',
                        place: 'Pittsburgh',
                        name: 'Temper',
                        teamname: 'temper',
                        seasonStart: '2022',
                        seasonEnd: '2022',
                    },
                } as SavedServerActionData)}
                teamOne={{ name: 'Team 1' }}
                teamTwo={{ name: 'Team 2' }}
                onPress={onPress}
            />,
        )

        const tag = getByText('break')
        fireEvent.press(tag)
        expect(onPress).toHaveBeenCalled()
    })
})
