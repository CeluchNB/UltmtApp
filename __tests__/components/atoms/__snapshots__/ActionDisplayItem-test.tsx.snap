// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot with live action 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#1a1a1a",
      "borderRadius": 8,
      "padding": 10,
      "shadowColor": "#ebebeb",
      "shadowRadius": 8,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffd600",
        "fontSize": 30.821917808219176,
        "fontWeight": "900",
      }
    }
  >
    Team 2
  </Text>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 30.821917808219176,
                "fontWeight": "bold",
              }
            }
          >
            Noah Celuch
          </Text>
        </View>
        <View>
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 30.821917808219176,
                "fontWeight": "900",
              }
            }
          >
             throws to 
          </Text>
        </View>
        <View>
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 30.821917808219176,
                "fontWeight": "bold",
              }
            }
          >
            Connor Tipping
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "#303030",
              "borderColor": "rgba(255, 214, 0, 0.29)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "elevation": 0,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": true,
                "selected": false,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Array [
                  Object {
                    "borderRadius": 4,
                  },
                  Object {
                    "flexGrow": 1,
                  },
                ],
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "paddingLeft": 4,
                    "position": "relative",
                  },
                  Object {
                    "paddingRight": 4,
                  },
                ]
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "lineHeight": 24,
                        "marginVertical": 4,
                        "minHeight": 24,
                        "textAlignVertical": "center",
                      },
                      Object {
                        "color": "rgba(255, 214, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "marginLeft": 8,
                        "marginRight": 8,
                      },
                      undefined,
                    ],
                  ]
                }
              >
                break
              </Text>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "#303030",
              "borderColor": "rgba(255, 214, 0, 0.29)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "elevation": 0,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": true,
                "selected": false,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Array [
                  Object {
                    "borderRadius": 4,
                  },
                  Object {
                    "flexGrow": 1,
                  },
                ],
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "paddingLeft": 4,
                    "position": "relative",
                  },
                  Object {
                    "paddingRight": 4,
                  },
                ]
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "lineHeight": 24,
                        "marginVertical": 4,
                        "minHeight": 24,
                        "textAlignVertical": "center",
                      },
                      Object {
                        "color": "rgba(255, 214, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "marginLeft": 8,
                        "marginRight": 8,
                      },
                      undefined,
                    ],
                  ]
                }
              >
                huck
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "margin": 6,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 11.25,
                  "height": 22.5,
                  "width": 22.5,
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <View>
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#ffd600",
                    "fontSize": 15,
                  },
                  Array [
                    Object {
                      "lineHeight": 15,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰆉
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`should match snapshot with server action 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#1a1a1a",
      "borderRadius": 8,
      "padding": 10,
      "shadowColor": "#ebebeb",
      "shadowRadius": 8,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffd600",
        "fontSize": 30.821917808219176,
        "fontWeight": "900",
      }
    }
  >
    Temper
  </Text>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 30.821917808219176,
                "fontWeight": "bold",
              }
            }
          >
            Noah Celuch
          </Text>
        </View>
        <View>
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 30.821917808219176,
                "fontWeight": "900",
              }
            }
          >
             throws to 
          </Text>
        </View>
        <View>
          <Text
            style={
              Object {
                "color": "#ffd600",
                "fontSize": 30.821917808219176,
                "fontWeight": "bold",
              }
            }
          >
            Connor Tipping
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "flexWrap": "wrap",
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "#303030",
              "borderColor": "rgba(255, 214, 0, 0.29)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "elevation": 0,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": true,
                "selected": false,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Array [
                  Object {
                    "borderRadius": 4,
                  },
                  Object {
                    "flexGrow": 1,
                  },
                ],
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "paddingLeft": 4,
                    "position": "relative",
                  },
                  Object {
                    "paddingRight": 4,
                  },
                ]
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "lineHeight": 24,
                        "marginVertical": 4,
                        "minHeight": 24,
                        "textAlignVertical": "center",
                      },
                      Object {
                        "color": "rgba(255, 214, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "marginLeft": 8,
                        "marginRight": 8,
                      },
                      undefined,
                    ],
                  ]
                }
              >
                break
              </Text>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "#303030",
              "borderColor": "rgba(255, 214, 0, 0.29)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "elevation": 0,
              "flexDirection": "column",
              "margin": 2,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": true,
                "selected": false,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Array [
                  Object {
                    "borderRadius": 4,
                  },
                  Object {
                    "flexGrow": 1,
                  },
                ],
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "flexGrow": 1,
                    "paddingLeft": 4,
                    "position": "relative",
                  },
                  Object {
                    "paddingRight": 4,
                  },
                ]
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "lineHeight": 24,
                        "marginVertical": 4,
                        "minHeight": 24,
                        "textAlignVertical": "center",
                      },
                      Object {
                        "color": "rgba(255, 214, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "marginLeft": 8,
                        "marginRight": 8,
                      },
                      undefined,
                    ],
                  ]
                }
              >
                huck
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 6,
              "left": 6,
              "right": 6,
              "top": 6,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "margin": 6,
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 11.25,
                  "height": 22.5,
                  "width": 22.5,
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          <View>
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#ffd600",
                    "fontSize": 15,
                  },
                  Array [
                    Object {
                      "lineHeight": 15,
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰆉
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
